modules/tensor_mechanics/src/actions/TensorMechanicsAction.C

```diff
@@ -40,45 +34,35 @@ void
 TensorMechanicsAction::act()
 {
   std::vector<NonlinearVariableName> displacements;
-  //This code will be the up-to-date version using the displacement string
   if (isParamValid("displacements"))
     displacements = getParam<std::vector<NonlinearVariableName> > ("displacements");
-
-  //This code is for the depricated version that allows three unique displacement variables
-  else if (isParamValid("disp_x"))
-  {
-    mooseDeprecated("StressDivergenceTensors has been updated to accept a string of displacement variable names, e.g. displacements = 'disp_x disp_y' in the input file.");
-    displacements.push_back(getParam<NonlinearVariableName>("disp_x"));
-    if (isParamValid("disp_y"))
-    {
-      displacements.push_back(getParam<NonlinearVariableName>("disp_y"));
-      if (isParamValid("disp_z"))
-         displacements.push_back(getParam<NonlinearVariableName>("disp_z"));
-    }
-  }
-
   else
     mooseError("The input file should specify a string of displacement names; these names should match the Variable block names.");
 
-  std::vector<VariableName> coupled_displacements;  //vector to hold the string of coupled disp variables
-  unsigned int dim = displacements.size();
-
-  for (unsigned int i = 0; i < dim; ++i)
+  unsigned int _ndisp = displacements.size();
+  std::vector<VariableName> coupled_displacements;
+  for (unsigned int i = 0; i < _ndisp; ++i)
     coupled_displacements.push_back(displacements[i]);
 
+  std::vector<std::vector<AuxVariableName> > save_in(_ndisp);
+  if (isParamValid("save_in"))
+  {
+    std::vector<AuxVariableName> this_save_in = getParam<std::vector<AuxVariableName> >("save_in");
+    if (this_save_in.size() != _ndisp)
+      mooseError("Number of save_in variables should equal to the number of displacement variables " << _ndisp);
+    for (unsigned int i = 0; i < _ndisp; ++i)
+      save_in[i].push_back(this_save_in[i]);
+  }
 
-  // Retain this code 'as is' because StressDivergenceTensors inherits from Kernel.C
-  std::vector<std::vector<AuxVariableName> > save_in(dim);
-
-  if (isParamValid("save_in_disp_x"))
-    save_in[0] = getParam<std::vector<AuxVariableName> >("save_in_disp_x");
-
-  if (isParamValid("save_in_disp_y"))
-    save_in[1] = getParam<std::vector<AuxVariableName> >("save_in_disp_y");
-
-  if (isParamValid("save_in_disp_z"))
-    save_in[2] = getParam<std::vector<AuxVariableName> >("save_in_disp_z");
-
+  std::vector<std::vector<AuxVariableName> > diag_save_in(_ndisp);
+  if (isParamValid("diag_save_in"))
+  {
+    std::vector<AuxVariableName> this_diag_save_in = getParam<std::vector<AuxVariableName> >("diag_save_in");
+    if (this_diag_save_in.size() != _ndisp)
+      mooseError("Number of diag_save_in variables should equal to the number of displacement variables " << _ndisp);
+    for (unsigned int i = 0; i < _ndisp; ++i)
+      diag_save_in[i].push_back(this_diag_save_in[i]);
```

Comment:
This logic seems overly complicated. Just get the parameter directly into `diag_save_in` and do the size check on that variable.